File: lib/src/parser/system_layout_parser.dart
import 'package:musicxml_parser/src/models/system_layout.dart';
import 'package:musicxml_parser/src/parser/xml_helper.dart';
import 'package:xml/xml.dart';

/// Parser for <system-layout> elements.
class SystemLayoutParser {
  /// Parses an [XmlElement] into a [SystemLayout] object.
  SystemLayout parse(XmlElement element) {
    // TODO: Implement parsing logic
    // Placeholder implementation
    SystemMargins? margins;
    final marginsElement = element.findElements('system-margins').firstOrNull;
    if (marginsElement != null) {
      margins = SystemMargins(
        leftMargin: XmlHelper.getElementTextAsDouble(marginsElement.findElements('left-margin').firstOrNull),
        rightMargin: XmlHelper.getElementTextAsDouble(marginsElement.findElements('right-margin').firstOrNull),
      );
    }

    SystemDividers? dividers;
    final dividersElement = element.findElements('system-dividers').firstOrNull;
    if (dividersElement != null) {
      dividers = SystemDividers(
        leftDivider: XmlHelper.getElementTextAsBool(dividersElement.findElements('left-divider').firstOrNull),
        rightDivider: XmlHelper.getElementTextAsBool(dividersElement.findElements('right-divider').firstOrNull),
      );
    }

    return SystemLayout(
      systemMargins: margins,
      systemDistance: XmlHelper.getElementTextAsDouble(element.findElements('system-distance').firstOrNull),
      topSystemDistance: XmlHelper.getElementTextAsDouble(element.findElements('top-system-distance').firstOrNull),
      systemDividers: dividers,
    );
  }
}

File: csharp/MusicXMLParser/Parser/SystemLayoutParser.cs
// Assuming necessary using statements for MusicXML models and helpers
using System.Xml.Linq;
using System.Linq;
using MusicXMLParser.Models; // For SystemLayout, SystemMargins, SystemDividers
using MusicXMLParser.Utils; // For XmlHelper

namespace MusicXMLParser.Parser
{
    /// <summary>
    /// Parser for <system-layout> elements.
    /// </summary>
    public class SystemLayoutParser
    {
        public SystemLayout Parse(XElement element)
        {
            SystemMargins margins = null;
            var marginsElement = element.Elements("system-margins").FirstOrDefault();
            if (marginsElement != null)
            {
                margins = new SystemMargins(
                    leftMargin: MusicXMLParser.Utils.XmlHelper.GetElementTextAsDouble(marginsElement.Elements("left-margin").FirstOrDefault()),
                    rightMargin: MusicXMLParser.Utils.XmlHelper.GetElementTextAsDouble(marginsElement.Elements("right-margin").FirstOrDefault())
                );
            }

            SystemDividers dividers = null;
            var dividersElement = element.Elements("system-dividers").FirstOrDefault();
            if (dividersElement != null)
            {
                // Assuming GetElementTextAsBool returns bool? to handle missing elements.
                // If it returns bool and defaults to false for missing, that's also fine.
                // The SystemDividers model should clarify if these are nullable or default to false.
                // Based on Dart's XmlHelper.getElementTextAsBool, it likely returns bool and defaults to false.
                dividers = new SystemDividers(
                    leftDivider: XmlHelper.GetElementTextAsBool(dividersElement.Elements("left-divider").FirstOrDefault()),
                    rightDivider: XmlHelper.GetElementTextAsBool(dividersElement.Elements("right-divider").FirstOrDefault())
                );
            }

            var systemDistance = XmlHelper.GetElementTextAsDouble(element.Elements("system-distance").FirstOrDefault());
            var topSystemDistance = XmlHelper.GetElementTextAsDouble(element.Elements("top-system-distance").FirstOrDefault());

            // Assuming SystemLayout constructor handles nullable properties appropriately.
            return new SystemLayout(
                systemMargins: margins,
                systemDistance: systemDistance,
                topSystemDistance: topSystemDistance,
                systemDividers: dividers
            );
        }
    }
}
